### PR TITLE
K8SPG-330: Preserve TLS secrets on cluster deletion and add finalizer to optionally delete them

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPG-330](https://jira.percona.com/browse/K8SPG-330) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

**CHANGE DESCRIPTION**
---
**Problem:**
We should preserve TLS secrets on cluster deletion to be consistent with our other operators.

**Cause:**
Crunchy controller set controller reference to TLS secrets for k8s to garbage collect them.

**Solution:**
Patch Crunchy controller to remove controller reference and add finalizer to delete TLS secrets.

**CHECKLIST**
---
**Jira**
- [x] Is the Jira ticket created and referenced properly?
- [x] Does the Jira ticket have the proper statuses for documentation (`Needs Doc`) and QA (`Needs QA`)?
- [x] Does the Jira ticket link to the proper milestone (Fix Version field)?

**Tests**
- [x] Is an E2E test/test case added for the new feature/change?
- [x] Are unit tests added where appropriate?

**Config/Logging/Testability**
- [x] Are all needed new/changed options added to default YAML files?
- [x] Are the manifests (crd/bundle) regenerated if needed?
- [x] Did we add proper logging messages for operator actions?
- [x] Did we ensure compatibility with the previous version or cluster upgrade process?
- [x] Does the change support oldest and newest supported PG version?
- [x] Does the change support oldest and newest supported Kubernetes version?

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/8ad36fb0a/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/8ad36fb0a/demand-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/8ad36fb0a/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/8ad36fb0a/monitoring.log) |
| We run 4 out of 4|

commit: https://github.com/percona/percona-postgresql-operator/pull/408/commits/8ad36fb0a9582802afcca2879a528efa471ae81b
image: `perconalab/percona-postgresql-operator:PR-408-8ad36fb0a`

File: percona/controller/pgcluster/version.go [hunk 1]
```diff
@@ -48,25 +48,31 @@ func (r *PGClusterReconciler) getVersionMeta(cr *v2beta1.PerconaPGCluster, opera
...
 	}
+
 	return vm
 }
 
 func (r *PGClusterReconciler) getOperatorDeployment(ctx context.Context) (*appsv1.Deployment, error) {
 	ns, err := k8s.GetOperatorNamespace()
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get operator namespace")
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (egegunes) [reply percona/controller/pgcluster/version.go:78]: fixed by ignoring this line. because not returning the error was on purpose

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/05fe9ef8c/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/05fe9ef8c/demand-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/05fe9ef8c/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/05fe9ef8c/monitoring.log) |
| We run 4 out of 4|

commit: https://github.com/percona/percona-postgresql-operator/pull/408/commits/05fe9ef8cf1522331f3c104c8e3ce96cd8ac835c
image: `perconalab/percona-postgresql-operator:PR-408-05fe9ef8c`

File: deploy/cr.yaml [hunk 2]
```diff
@@ -4,6 +4,7 @@ metadata:
   name: cluster1
 #  finalizers:
 #  - percona.com/delete-pvc
+#  - percona.com/delete-tls
```
Comment (inelpandzic) [inline deploy/cr.yaml:7]: In our other operators this is named `delete-ssl`, we should make it the same here as well?

Comment (egegunes) [reply deploy/cr.yaml]: you're right, switched to `delete-ssl`

File: pkg/apis/pg.percona.com/v2beta1/perconapgcluster_types.go [hunk 3]
```diff
@@ -514,4 +514,5 @@ func GetDefaultVersionServiceEndpoint() string {
 
 const (
 	FinalizerDeletePVC = "percona.com/delete-pvc"
+	FinalizerDeleteTLS = "percona.com/delete-ssl"
```
Comment (inelpandzic) [inline pkg/apis/pg.percona.com/v2beta1/perconapgcluster_types.go:517]:
```suggestion
	FinalizerDeleteSSL = "percona.com/delete-ssl"
```

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/b49bc7263/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/b49bc7263/demand-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/b49bc7263/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/b49bc7263/monitoring.log) |
| We run 4 out of 4|

commit: https://github.com/percona/percona-postgresql-operator/pull/408/commits/b49bc726309d49a36c6c9ddac56cb63f5e9adacd
image: `perconalab/percona-postgresql-operator:PR-408-b49bc7263`

File: pkg/apis/pg.percona.com/v2beta1/perconapgcluster_types.go (hunk 3 above)
Comment (egegunes) [reply pkg/apis/pg.percona.com/v2beta1/perconapgcluster_types.go]: fixed

**Review by inelpandzic — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/e7a3eca93/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/e7a3eca93/demand-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/e7a3eca93/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/e7a3eca93/monitoring.log) |
| We run 4 out of 4|

commit: https://github.com/percona/percona-postgresql-operator/pull/408/commits/e7a3eca93d8056f7745273a81c16e5485944ade6
image: `perconalab/percona-postgresql-operator:PR-408-e7a3eca93`

**Review by inelpandzic — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/4a6ffed67/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/4a6ffed67/demand-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/4a6ffed67/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-408/4a6ffed67/monitoring.log) |
| We run 4 out of 4|

commit: https://github.com/percona/percona-postgresql-operator/pull/408/commits/4a6ffed6709cd81c62fcefef2d0c524397724945
image: `perconalab/percona-postgresql-operator:PR-408-4a6ffed67`

**Review by hors — APPROVED** [review]

**Pull request merged**